core/src/main/java/io/parsingdata/metal/data/DataExpressionSource.java

```diff
@@ -48,28 +51,32 @@ public DataExpressionSource(final ValueExpression dataExpression, final int inde
     @Override
     protected byte[] getData(final BigInteger offset, final BigInteger length) {
         checkNotNegative(offset, "offset");
-        final Value inputValue = getValue();
-        if (checkNotNegative(length, "length").add(offset).compareTo(inputValue.slice.length) > 0) {
+        final byte[] data = getValue();
+        if (checkNotNegative(length, "length").add(offset).compareTo(BigInteger.valueOf(data.length)) > 0) {
             throw new IllegalStateException("Data to read is not available ([offset=" + offset + ";length=" + length + ";source=" + this + ").");
         }
         final byte[] outputData = new byte[length.intValueExact()];
-        System.arraycopy(inputValue.getValue(), offset.intValueExact(), outputData, 0, outputData.length);
+        System.arraycopy(data, offset.intValueExact(), outputData, 0, outputData.length);
         return outputData;
     }
 
     @Override
     protected boolean isAvailable(final BigInteger offset, final BigInteger length) {
-        return checkNotNegative(offset, "offset").add(checkNotNegative(length, "length")).compareTo(getValue().slice.length) <= 0;
+        return checkNotNegative(offset, "offset").add(checkNotNegative(length, "length")).compareTo(BigInteger.valueOf(getValue().length)) <= 0;
     }
 
-    private Value getValue() {
-        final ImmutableList<Optional<Value>> results = dataExpression.eval(parseState, encoding);
-        if (results.size <= index) {
-            throw new IllegalStateException("ValueExpression dataExpression yields " + results.size + " result(s) (expected at least " + (index + 1) + ").");
+    private synchronized byte[] getValue() {
```

Comment:
No double checked locking to avoid useless synchronization after it has been cached? :smile:

Reply:
I just read the pr comment, so okay :smile: